(DOCS-3043) how appsec works guide (#13053)
DataDog/documentation: content/en/security_platform/guide/how-appsec-works.md

---
title: How Application Security works in Datadog
kind: guide
---

> ## Overview

Reviewer: Worth highlighting it relies on the same Library than is used by the APM suite, thus not requiring teams to deploy and maintain yet another one

Application Security monitors application-level attacks aiming to exploit code-level vulnerabilities.

Application Security records information about each request and applies security rules in-app. This information enriches your APM traces with security data.

Traditional Web Application Firewalls (WAFs) are usually deployed at the perimeter and have no context of the application behavior. Application Security provides WAF-like capabilities with additional application context and lower false positives.

## Threat detection methods

Datadog uses multiple pattern sources, including the [OWASP ModSecurity Core Rule Set][1] to detect known threats and vulnerabilities in HTTP requests. When an HTTP request matches one of those rules, a security signal is generated in Datadog.

> Signals are automatically created when Datadog detects meaningful attacks targeting your production services. It provides you with visibility on the attackers and the targeted services. You can set monitors with thresholds to determine which attacks you want to be notified about.

Reviewer: Security Signals

Reviewer: Monitors > Custom Detection rules

> ## Attack types

Reviewer: Coverage

> Application Security includes over 100 OOTB detection rules that help safegaurd against many different kinds of attacks, including the following:

Reviewer: safeguard?

* SQL injections
* Code Injections 
* Shell Injections
* NoSQL injections 
* Cross-Site Scripting (XSS) 

> * Sever-side Request Forgery (SSRF)

Reviewer: Full source, see the type facet in the AppSec configuration page https://app.datadoghq.com/security/appsec/event-rules

Feel free to link to this

> Application Security categorizes these attacks into different threat types and notifies Datadog services such as monitors and events, based on your settings.

Reviewer: Not sure what this paragraph aims to describe

* **Unqualified attacks** match inbound HTTP requests with known attack patterns, but don't have any correlation to the business-logic.
* **Contextualized attacks** have a correlation between known attack patterns and business-logic.

> * A **Vulnerability is triggered** when there is a correlation between attack patterns, and there's evidence of a successful vulnerability trigger.

Reviewer: Should go above imho

---

> title: How Application Security works in Datadog

Reviewer 2: ```suggestion
title: How Application Security Works in Datadog
```

kind: guide
---

## Overview

Datadog Application Security monitors application-level attacks aiming to exploit code-level vulnerabilities.

> APM records information about each HTTP request, these are referred to as traces. Datadog Application Security uses the information APM is already collecting, and flags any attack attempts based or suspicious requests that match known attack patterns. Security signals are an aggregation of suspicious requests. Depending on your security signal settings, you can receive notifications from Slack, email, or PagerDuty.

Reviewer 2: ```suggestion
APM records information about each HTTP request, referred to as traces. Datadog Application Security uses the information APM is already collecting, and flags attack attempts based on suspicious requests that match known attack patterns. Security signals are an aggregation of suspicious requests. Depending on your security signal settings, you can receive notifications from Slack, email, or PagerDuty.
```

> Traditional Web Application Firewalls (WAFs) are usually deployed at the perimeter and have no context of the application behavior. For application security to be effective, it’s critical to be embedded in the application to get access to the data. Datadog Application Security leverages known attack patterns, similar to a Web Application Firewall (WAF) but with additional application context to increase the signal to noise ratio, lowering false positives.

Reviewer 2: ```suggestion
Traditional Web Application Firewalls (WAFs) are usually deployed at the perimeter and have no context of the application behavior. For Application Security to be effective, it must be embedded in the application to get access to the data. Datadog Application Security leverages known attack patterns, similar to a Web Application Firewall (WAF) but with additional application context to increase the signal to noise ratio, lowering false positives.
```

### Compatibility

> For Datadog Application Security to be compatible with your Datadog configuration, you must have APM enabled, and [send traces to Datadog][1]. It uses the same libraries used by the APM suite, so you don't need to deploy and maintain another library for this task. Steps to enable Datadog Application Security are specific to runtime language. Check to see if your language is supported in the [Application Security prerequisites][2].

Reviewer 2: ```suggestion
For Datadog Application Security to be compatible with your Datadog configuration, you must have APM enabled, and [send traces to Datadog][1]. Application Security uses the same libraries used by APM, so you don't need to deploy and maintain another library. Steps to enable Datadog Application Security are specific to runtime language. Check to see if your language is supported in the [Application Security prerequisites][2].
```

### Performance

> Datadog Application Security uses processes already contained in the Agent and APM, so there are no performance implications when using it. Datadog Application Security collects and enriches trace data from APM. It doesn't require extra overhead for processes like analyzing attack attempts, because it matches against known patterns and flags them. All processing is done platform-side, after traces have been flagged. 

Reviewer 2: ```suggestion
Datadog Application Security uses processes already contained in the Agent and APM, so there are no performance implications when using it. When APM is enabled, the Datadog Library generates distributed traces. Datadog Application Security flags security activity in traces by using known attack patterns. Correlation between the attack patterns and the execution context provided by the distributed trace triggers security signals based on detection rules.
```

### Data privacy

There are multiple methods used to avoid your sensitive information being indexed. To take further action, you can set up [custom and static scrubbers][3], and use [exclusion filters][4].

**Note:** Datadog Application Security does not automatically obfuscate sensitive information or PII. To keep this sensitive data from being sent to Datadog, [configure the Datadog Agent or Tracer for data security][3].

Contact Support to delete sensitive data that may have been indexed.

## Threat detection methods

Datadog uses multiple pattern sources, including the [OWASP ModSecurity Core Rule Set][6] to detect known threats and vulnerabilities in HTTP requests. When an HTTP request matches one of [the OOTB detection rules][7], a security signal is generated in Datadog.

Security Signals are automatically created when Datadog detects meaningful attacks targeting your production services. It provides you with visibility on the attackers and the targeted services. You can set custom detection rules with thresholds to determine which attacks you want to be notified about.

## Coverage

Datadog Application Security categorizes attack attempts into different threat types:

> * **Unqualified attacks** match inbound HTTP requests with known attack patterns, but don't have any correlation to the business-logic.

Reviewer 2: ```suggestion
* **Unqualified attacks** match inbound HTTP requests with known attack patterns. For example, no correlation with the service's business-logic is found after correlating with the execution context provided by the trace.
```

> * **Contextualized attacks** have a correlation between known attack patterns and business-logic.

Reviewer 2: ```suggestion
* **Contextualized attacks** correlate the attack attempts performed on the service with a matching business-logic. For example, SQL injection patterns on a service performing SQL statements.
```

---
title: How Application Security works in Datadog
kind: guide
---

## Overview

Application Security monitors application-level attacks aiming to exploit code-level vulnerabilities.

> APM records information about each HTTP request, these are referred to as traces. Application security uses the information APM is already collecting, and flags any suspicious requests or attack attempts as security signals. Depending on your settings and the severity of the attempt, you can receive notifications from Slack, email, or PagerDuty.

Reviewer: attack attempt is what we detect. "suspicious" is how Application Security qualifies a request where such attempts has been detected

Reviewer: > Depending on your settings and the severity of the attempt, you can receive notifications from Slack, email, or PagerDuty.

Not attack attempt level, only Security Signals can alert, as they aggregate multiple traces and detect attacks impacting production systems

> APM records information about each HTTP request, these are referred to as traces. Datadog Application Security uses the information APM is already collecting, and flags any attack attempts based on suspicious requests as security signals. Depending on your security signal settings, you can receive notifications from Slack, email, or PagerDuty.

Reviewer: > Datadog Application Security uses the information APM is already collecting, and flags any attack attempts based on suspicious requests as security signals

Still not accurate as a Security Signals is an aggregation of suspicious requests.

A Suspicious requests is a trace where security activity - today attack attempts - have been flagged.

> APM records information about each HTTP request, these are referred to as traces. Datadog Application Security uses the information APM is already collecting, and flags any attack attempts based on suspicious requests as security signals. Depending on your security signal settings, you can receive notifications from Slack, email, or PagerDuty.

Author: ```suggestion
APM records information about each HTTP request, these are referred to as traces. Datadog Application Security uses the information APM is already collecting, and flags any attack attempts based or suspicious requests that match known attack patterns. Security signals are an aggregation of suspicious requests. Depending on your security signal settings, you can receive notifications from Slack, email, or PagerDuty.
```

> Traditional Web Application Firewalls (WAFs) are usually deployed at the perimeter and have no context of the application behavior. For Application Security to be effective, it’s critical to be embedded in the application to get access to the data. Application Security provides WAF-like capabilities with additional application context and lower false positives.

Reviewer: > For Application Security to be effective
Confusing if about the domain or Datadog product. Maybe use lower-case and always refer to the product as "Datadog Application Security"?

> Application Security provides WAF-like capabilities with additional application context and lower false positives.
Application Security leverages known attack patterns like WAF and leverage the execution context to increase the signal/noise ratio, i.e reduce false positives

### Compatibility

For Application Security to be compatible with your Datadog configuration, you must have APM enabled, and [send traces to Datadog][1]. It uses the same libraries used by the APM suite, so you don't need to deploy and maintain another library for this task. Steps to enable Application Security are specific to runtime language. Check to see if your language is supported in the [Application Security prerequisites][2].

> ### Performance

Reviewer: Good to mention that the additional processing is done platform-wise. All what is done on the service-side is matching the attack patterns

Application Security uses processes already contained in the Agent and APM, so there are no performance implications when using it. Application Security collects and enriches trace data from APM. It doesn't require extra overhead for processes like analyzing attack attempts, because it matches against known patterns and flags them.

> ### Data Privacy

Author: ```suggestion
### Data privacy
```

> Application Security only scans HTTP request headers, it does not scan the body. 

Reviewer: Not true anymore, some libraries do scan the body now

> Datadog Application Security uses processes already contained in the Agent and APM, so there are no performance implications when using it. Datadog Application Security collects and enriches trace data from APM. It doesn't require extra overhead for processes like analyzing attack attempts, because it matches against known patterns and flags them. All processing is done platform-side, after traces have been flagged. 

Reviewer: > Datadog Application Security collects and enriches trace data from APM. It doesn't require extra overhead for processes like analyzing attack attempts, because it matches against known patterns and flags them. All processing is done platform-side, after traces have been flagged.

- When APM is enabled, the Datadog Library generates Distributed Traces
- Datadog Application Security flags security activity in traces - today attack attempts using known attack patterns. This adds no overhead and no traffic redirection
- Then, platform-wise, the correlation between the attack patterns and the execution context provided by the distributed trace is performed enabling to detect security signals based on detection rules (OOTB, custom)

There are multiple methods used to avoid your sensitive information being indexed. To take further action, you can set up [custom and static scrubbers][3], and use [exclusion filters][4].

> The [Sensitive Data Scanner][5] automatically scans traces that have been flagged by Datadog Application Security.

Reviewer: Double-checked with the team. Not released yet

> The [Sensitive Data Scanner][5] automatically scans traces that have been flagged by Datadog Application Security.

Author: ```suggestion
```

The [Sensitive Data Scanner][5] automatically scans traces that have been flagged by Application Security.

Application Security does not automatically obfuscate sensitive information or PII. To keep this sensitive data from being sent to Datadog, [configure the Datadog Agent or Tracer for data security][3].

Contact Support to delete sensitive data that may have been indexed.

## Threat detection methods

> Datadog uses multiple pattern sources, including the [OWASP ModSecurity Core Rule Set][6] to detect known threats and vulnerabilities in HTTP requests. When an HTTP request matches one of those rules, a security signal is generated in Datadog.

Reviewer: > When an HTTP request matches one of those rules, a security signal is generated in Datadog.
Only when one of the OOTB detection rules match. Otherwise it'd be too noisy

> Security Signals are automatically created when Datadog detects meaningful attacks targeting your production services. It provides you with visibility on the attackers and the targeted services. You can set custom detection rules with thresholds to determine which attacks you want to be notified about.

Reviewer: Link to OOTB rules for Application Security

## Coverage

Datadog Application Security categorizes attack attempts into different threat types:

> * **Unqualified attacks** match inbound HTTP requests with known attack patterns, but don't have any correlation to the business-logic.

Reviewer: * **Unqualified attacks** match inbound HTTP requests with known attack patterns. No correlation with the service's business-logic was found after correlating with the execution context provided by the trace

> * **Contextualized attacks** have a correlation between known attack patterns and business-logic.

Reviewer: * **Contextualized attacks** the attack attempts performed on the service was correlated with a matching business-logic. For example, SQL injection patterns on a service performing SQL statements

* **Contextualized attacks** have a correlation between known attack patterns and business-logic.
* A **Vulnerability is triggered** when there is a correlation between attack patterns, and there's evidence of a successful vulnerability trigger.

> Application Security includes over 100 OOTB detection rules that help protect against [many different kinds of attacks][7], including the following:

Reviewer: 100 attack patterns. Detection rule is another thing

> * A **Vulnerability is triggered** when there is a correlation between attack patterns, and there's evidence of a successful vulnerability trigger.

Reviewer: * **Vulnerability is triggered** evidences of a vulnerability successfully being exploited have been found after matching known attack patterns

> Datadog Application Security includes over 100 attack patterns that help protect against [many different kinds of attacks][8], including the following:

Reviewer: Datadog Application Security includes over 100 attack patterns that help protect against [many different kinds of attacks][8], covering the following _vulnerabilities_:

> Datadog Application Security includes over 100 attack patterns that help protect against [many different kinds of attacks][8], including the following:

Author: ```suggestion
Datadog Application Security includes over 100 attack patterns that help protect against [many different kinds of attacks][8], including the following vulnerabilities:
```

* SQL injections
* Code injections
* Shell injections
* NoSQL injections
* Cross-Site Scripting (XSS)
* Sever-side Request Forgery (SSRF)

### How Application Security protects against Log4Shell

>  Application Security identifies Log4Shell attack payloads and provides visibility into vulnerable apps that attempt to remotely load malicious code. When used in tandem with the rest of [Datadog's Security Platform][8], you can investigate to identify common post-exploitation activity, and proactively remediate potentially exposed Java applications acting as an attack vector.

Reviewer: > Log4Shell attack payloads
Log4j Log4Shell attack payloads

> visibility into vulnerable apps
vulnerable Java web services

> When used in tandem with the rest of [Datadog's Security Platform][8]
More specifically Datadog Cloud Workload Security here
